Sqlalchemy

nickeopti opened the pull request:
~~Draft, WIP, PR. Just made to keep track of things, and for the tests to run. Can be ignored for now.~~

Edit: PR for improving the structure and codebase of the drivers. Was originally just for replacing the database specific code with a more general SQLAlchemy implementation. But has now also turned into a bigger, more general refactor of the entire driver code.

nickeopti commented:
This refactor is now functionally near-identical to the `main` branch. Along the way, I have also taken the liberty to simplify and improve readability of some things.

I think this version is significantly more readable, and makes it easier to add new functionality -- both adding new methods and new drivers will require less code, as the new `RelationalDriver` handles most things centrally, such that subclasses (such as `MySQLDriver` and `SQLiteDriver`) just has to specify relatively few database-vendor specific things.

Additionally, I have begun to prefer the style of creating of SQL queries through SQLAlchemy Core, rather than crafting them by combining string snippets. I especially like how e.g. `WHERE` clauses can be built by splatting a list (comprehension) of conditions -- which may evaluate to the empty list when no conditions are supplied, in which case the `WHERE` clause is simply omitted. Same goes for parenthesized `OR` conditions. 

All major functionality is preserved -- however, there are some smaller functional changes:
1. Previous `sqlite` driver checked DB version compatibility every time a connection was established, while `mysql` driver only checked once per driver instance. This branch uses the `mysql` driver approach for all drivers.
2. The drivers' `_connection_callback` is renamed to more descriptive `_verify_db_version`. This is slightly weird in the `sqlite_remote` driver, which actually uses that method as a generic callback, but I still believe it is more descriptive.
3. Several internal instance variables are renamed as well. If this may be troublesome for users depending on them, then lets rename them back again. But it is only done on instance variables, so I kinda hope no one relies on them.
4. Previous `mysql` driver cached keys, while the `sqlite` didn't. This branch uses `mysql` approach for all drivers.
5. `settings.LAZY_LOADING_MAX_SHAPE` was used when lazy loading metadata in the previous `sqlite` driver, but not in the `mysql` driver. For simplicity, this branch uses the `mysql` approach (of ignoring `settings.LAZY_LOADING_MAX_SHAPE`) for all drivers. I honestly have no idea what that does, nor why it was implemented only in the one driver. Please tell me about it, if it is relevant :D 
6. `terracotta.get_driver()` now caches drivers only per process, not globally. This is a necessity due to how the SQLAlchemy engine is created on driver instance instantiation, whereas the previous code only created connections when they were about the be used. Otherwise the same engine could end up being shared across processes, which resulted in some curious errors.
7. The `key_names` table is created with an additional `index` column, which is used to ensure that keys are retrieved from the database in the correct order. This is vital, and currently a breaking change. We need to decide how to handle this. This references #241.
8. The `mysql` no longer has a `MySQLCredentials` object to store details. Instead, this information is used straight-away to create the connection string.
9. The `RemoteSQLiteDriver` now has a `_local_path` field, which corresponds to what was previously contained in the inherited `path` field. The `path` field now contains a connection string instead.

Also, this branch handles paths somewhat differently -- particularly internally. This is done in an attempt to make the handling of paths and particularly connection strings as database-vendor agnostic as possible. However, I've also tried to retain the current functionality and conventions, and this has honestly lead to the entire path/connection string/connection parameters handling being somewhat clumsy. It works, so I guess everything is acceptable, but I really think we could improve that part of the code base significantly, _if_ we are ready to change the assumptions on how paths are stored, accessed, and normalized -- stemming all the way from the generic `Driver` class. However, that may belong to a future PR.

I have broken _one_ connection string input pattern, though. Connection strings of the form `test.example.com:3306/foo` does not currently work. Paths like
- `test.example.com/foo`
- `user@test.example.com/foo`
- `user:password@test.example.com/foo`
- `mysql://test.example.com:3306/foo`
- `mysql://user:password@test.example.com/foo`
- `mysql://user:password@test.example.com:3306/foo`

all work, but specifying a `port` without specifying a `scheme` does not. In fact, this trips up `urllib.parse.urlparse`, which this branch relies upon. We can of course implement some hard-coded checks for that particular path pattern -- but if it is alright with all of you, I think I would actually like to disallow that path pattern: if it is not parsed correctly by `urlparse`, shouldn't we consider it an invalid path, rather than try to fix the path?
The newly added failing `tests/drivers/test_drivers.py::test_normalize_url[mysql]` test covers exactly that question.

I have no idea what causes the `setup.py::mypy-status` and `terracotta/cache.py::mypy` tests to fail -- nor am I sure whether it has anything to do with these code changes.

BTW, even though SQLAlchemy supports thread-safe connection pools, this current implementation is _not_ thread-safe. As far as I can figure out, we would have to rewrite how the `requires_connection` decorator works quite substantially in order to achieve thread-safety. I propose this belongs to a future PR, if it is something we want to have.

dionhaefner commented on terracotta/drivers/relational_base.py:
```diff
@@ -0,0 +1,486 @@
+
+    @classmethod
+    def _parse_connection_string(cls, connection_string: str) -> urlparse.ParseResult:
+        con_params = urlparse.urlparse(connection_string)
```
```suggestion
        if "//" not in connection_string:
            connection_string = f"//{connection_string}"

        con_params = urlparse.urlparse(connection_string)
```

This fixes your failing test and is still clean enough IMO.

nickeopti replied:
Ah, neat solution! This required one extra change a few lines below. I just committed that locally, as I couldn't figure out how to edit your suggestion here. Sorry for not giving you the credit in the logs

dionhaefner commented:
Thanks a lot @nickeopti, this looks very high quality on first glance. I will do a proper review soon.

dionhaefner commented:
> All major functionality is preserved -- however, there are some smaller functional changes:
> 
> 1. Previous `sqlite` driver checked DB version compatibility every time a connection was established, while `mysql` driver only checked once per driver instance. This branch uses the `mysql` driver approach for all drivers.

OK

> 2. The drivers' `_connection_callback` is renamed to more descriptive `_verify_db_version`. This is slightly weird in the `sqlite_remote` driver, which actually uses that method as a generic callback, but I still believe it is more descriptive.

This is supposed to be a generic hook that just happens to be mostly about verifying the DB version. I prefer `_connection_callback`.
 
> 3. Several internal instance variables are renamed as well. If this may be troublesome for users depending on them, then lets rename them back again. But it is only done on instance variables, so I kinda hope no one relies on them.

This is fine.

> 4. Previous `mysql` driver cached keys, while the `sqlite` didn't. This branch uses `mysql` approach for all drivers.

I don't see any caching logic in any of the drivers?

> 5. `settings.LAZY_LOADING_MAX_SHAPE` was used when lazy loading metadata in the previous `sqlite` driver, but not in the `mysql` driver. For simplicity, this branch uses the `mysql` approach (of ignoring `settings.LAZY_LOADING_MAX_SHAPE`) for all drivers. I honestly have no idea what that does, nor why it was implemented only in the one driver. Please tell me about it, if it is relevant :D

Yes, this is very much relevant and should be used. The fact that this wasn't used in the MySQL driver was a bug.

Lazy loading means computing raster metadata the first time it is requested instead of during ingestion. But there's a dilemma: Computing metadata is costly, so if you have a huge raster the request would time out. Instead we compute the dataset on a smaller version of big rasters (downsampled to `LAZY_LOADING_MAX_SHAPE`) to trade accuracy for speed. So lazy loading is about quick-and-dirty data visualization, not production quality services.

> 6. `terracotta.get_driver()` now caches drivers only per process, not globally. This is a necessity due to how the SQLAlchemy engine is created on driver instance instantiation, whereas the previous code only created connections when they were about the be used. Otherwise the same engine could end up being shared across processes, which resulted in some curious errors.

OK

> 7. The `key_names` table is created with an additional `index` column, which is used to ensure that keys are retrieved from the database in the correct order. **This is vital, and currently a breaking change**. We need to decide how to handle this. This references [Internal reliance on physical order of DB rows #241](https://github.com/DHI-GRAS/terracotta/issues/241).

Ideally we would handle this via a migration script, see #245. If you're not tired of working on Terracotta yet it would be great if you could handle this :) Otherwise we'll have to accept that people will have to re-create their databases.

> 8. The `mysql` no longer has a `MySQLCredentials` object to store details. Instead, this information is used straight-away to create the connection string.

OK

> 9. The `RemoteSQLiteDriver` now has a `_local_path` field, which corresponds to what was previously contained in the inherited `path` field. The `path` field now contains a connection string instead.

OK

> Also, this branch handles paths somewhat differently -- particularly internally. This is done in an attempt to make the handling of paths and particularly connection strings as database-vendor agnostic as possible. However, I've also tried to retain the current functionality and conventions, and this has honestly lead to the entire path/connection string/connection parameters handling being somewhat clumsy. It works, so I guess everything is acceptable, but I really think we could improve that part of the code base significantly, _if_ we are ready to change the assumptions on how paths are stored, accessed, and normalized -- stemming all the way from the generic `Driver` class. However, that may belong to a future PR.

What specifically would you improve?

> I have broken _one_ connection string input pattern, though. Connection strings of the form `test.example.com:3306/foo` does not currently work. Paths like
> 
>     * `test.example.com/foo`
> 
>     * `user@test.example.com/foo`
> 
>     * `user:password@test.example.com/foo`
> 
>     * `mysql://test.example.com:3306/foo`
> 
>     * `mysql://user:password@test.example.com/foo`
> 
>     * `mysql://user:password@test.example.com:3306/foo`
> 
> 
> all work, but specifying a `port` without specifying a `scheme` does not. In fact, this trips up `urllib.parse.urlparse`, which this branch relies upon. We can of course implement some hard-coded checks for that particular path pattern -- but if it is alright with all of you, I think I would actually like to disallow that path pattern: if it is not parsed correctly by `urlparse`, shouldn't we consider it an invalid path, rather than try to fix the path? The failing `tests/drivers/test_drivers.py::test_normalize_url[mysql]` test covers exactly that question.

See suggestion above.
 
> I have no idea what causes the `setup.py::mypy-status` and `terracotta/cache.py::mypy` tests to fail -- nor am I sure whether it has anything to do with these code changes.

It has nothing to do with your changes, just mypy getting stricter over time.
 
> BTW, even though SQLAlchemy supports thread-safe connection pools, this current implementation is _not_ thread-safe. As far as I can figure out, we would have to rewrite how the `requires_connection` decorator works quite substantially in order to achieve thread-safety. I propose this belongs to a future PR, if it is something we want to have.

Ultimately we would definitely want thread-safety. I've opened an issue to track this (#246).

dionhaefner reviewed (changes requested):
Thanks again for the implementation!

I'm a bit torn regarding the new structure of the code (a lot of this is inherited from the old version, so not really your fault). Particularly, I think the inheritance tree is way too deep: `Driver -> RasterDriver -> RelationalDriver -> SQLiteDriver -> RemoteSQLiteDriver`. If we can I would like to stick to a maximum of 3 levels.

I think we should remove the `RasterDriver` mixin (#159), but that is something we can do after handling this PR. Right now, I see 2 ways going forward:

1. Remove some logic from `RelationalDriver` and move it to the subclasses to purge some class variables. Do not make `RemoteSQLiteDriver` inherit from `SQLiteDriver`, but from `RelationalDriver` instead. This comes at the cost of a bit more code duplication but should make the interface much cleaner.
2. Remove all subclasses (`SQLiteDriver`, `MySQLDriver`, `RemoteSQLiteDriver`) and handle everything inside `RelationalDriver`. I.e., `driver = RelationalDriver(provider="sqlite", path="..."). This comes at the cost of increased complexity and some extra conditionals in the  `RelationalDriver` but also eliminates more inheritance.

I am leaning towards option 1 because I like how clean we can implement e.g. the remote SQLite driver (just override some methods and you're done), but what is your opinion?

nickeopti commented on terracotta/drivers/mysql.py:
```diff
@@ -175,325 +67,14 @@ def _parse_db_name(con_params: ParseResult) -> str:
+        engine = sqla.create_engine(
+            f'{self._CONNECTION_PARAMETERS.scheme}+{self.SQL_DRIVER_TYPE}://'
+            f'{self._CONNECTION_PARAMETERS.netloc}',
+            echo=True,
```
The `echo`? Yeah, that was a leftover from testing. Nice spotted!

nickeopti commented on terracotta/drivers/mysql.py:
```diff
@@ -92,22 +27,12 @@ class MySQLDriver(RasterDriver):
-        ('percentiles', 'BLOB'),
-        ('metadata', 'LONGTEXT')
-    )
-    _CHARSET: str = 'utf8mb4'
```
It appears not. Running
```python
import terracotta
db = terracotta.get_driver('root:password@localhost/tc_test', provider='mysql')
db.create(['a', 'b'])
```
and then opening the `mysql` database and querying
```SQL
SELECT table_name, column_name, character_set_name FROM information_schema.`COLUMNS` WHERE table_schema = "tc_test";
```
yields
```
+------------+------------------+--------------------+
| TABLE_NAME | COLUMN_NAME      | CHARACTER_SET_NAME |
+------------+------------------+--------------------+
| datasets   | a                | utf8mb4            |
| datasets   | b                | utf8mb4            |
| datasets   | filepath         | utf8mb4            |
| key_names  | description      | utf8mb4            |
| key_names  | index            | NULL               |
| key_names  | key_name         | utf8mb4            |
| metadata   | a                | utf8mb4            |
| metadata   | b                | utf8mb4            |
| metadata   | bounds_east      | NULL               |
| metadata   | bounds_north     | NULL               |
| metadata   | bounds_south     | NULL               |
| metadata   | bounds_west      | NULL               |
| metadata   | convex_hull      | utf8mb4            |
| metadata   | max              | NULL               |
| metadata   | mean             | NULL               |
| metadata   | metadata         | utf8mb4            |
| metadata   | min              | NULL               |
| metadata   | percentiles      | NULL               |
| metadata   | stdev            | NULL               |
| metadata   | valid_percentage | NULL               |
| terracotta | version          | utf8mb4            |
+------------+------------------+--------------------+
```

However, I don't know why it is not (any longer?) necessary, nor whether it might be a good defensive strategy to use it.

dionhaefner replied:
Pretty sure this is MySQL vendor / version dependent. Better supply it to be safe.

nickeopti commented on terracotta/drivers/mysql.py:
```diff
@@ -175,325 +67,14 @@ def _parse_db_name(con_params: ParseResult) -> str:
-            read_timeout=self.DB_CONNECTION_TIMEOUT,
-            write_timeout=self.DB_CONNECTION_TIMEOUT,
-            binary_prefix=True, charset='utf8mb4'
+    def _create_database(self) -> None:
```
Perhaps we can. The rationale behind not doing it is that
- `sqlite` databases shall not be created; they are automatically created on first connection if they don't exist
- `mysql` databases do have to be created like this
- if we add a `postgres` driver: `postgres` databases also have to be created like this, however the connection string will have to be different, as `postgres` RDMS requires to be connected to an existing database (in which case we would use the default `postgres` database).

The issue regarding connection to already existing databases or not is also the reason that the `connect` method can't be used.

nickeopti commented on terracotta/drivers/relational_base.py:
```diff
@@ -0,0 +1,486 @@
+    try:
+        yield
+    except exceptions_to_convert as exception:
+        raise exceptions.InvalidDatabaseError(error_message) from exception
```
That is much nicer! Also a good point that `DATABASE_DRIVER_EXCEPTIONS_TO_CONVERT` no longer needs to be a class variable.

nickeopti commented on terracotta/drivers/relational_base.py:
```diff
@@ -0,0 +1,486 @@
+        raise exceptions.InvalidDatabaseError(error_message) from exception
+
+
+class RelationalDriver(RasterDriver, ABC):
```
Hmm, I thought that was a necessity... What is the alternative?

Well, fascinatingly, to me at least, this inheritance hierarchy 
```python
class RelationalDriver(Driver, ABC):
    ...
class MySQLDriver(RelationalDriver, RasterDriver):
    ...
```
actually works, though it seems really odd to me, as `RelationalDriver` needs `RasterDriver.compute_metadata`.

Was that what you had in mind, or is there some third alternative also?

dionhaefner replied:
Yea let's discuss first what we would like the class structure to be, then make a decision.

nickeopti commented on terracotta/drivers/relational_base.py:
```diff
@@ -0,0 +1,486 @@
+
+        if con_params.scheme == 'file' and cls.FILE_BASED_DATABASE:
+            file_connection_string = connection_string[len('file://'):]
+            con_params = urlparse.urlparse(f'{cls.SQL_DATABASE_SCHEME}://{file_connection_string}')
```
It's a work-around to getting file-paths (such as the ones used for the `SQLiteDriver`) parsed correctly by `urlparse`. Notice that `super().__init__(f'file:///{path}')` is called from the `SQLiteDriver` init.

I agree it's a bit weird, but it works, and I don't think it is too bad.

nickeopti commented on terracotta/drivers/relational_base.py:
```diff
@@ -0,0 +1,486 @@
+        assert self.SQL_DRIVER_TYPE is not None
+        self._CONNECTION_PARAMETERS = self._parse_connection_string(path)
+        cp = self._CONNECTION_PARAMETERS
+        resolved_path = self._resolve_path(cp.path[1:])  # Remove the leading '/'
```
Yes, by its current design, it will always be there. This is part of the reason for the 
```python
        if con_params.scheme == 'file' and cls.FILE_BASED_DATABASE:
            file_connection_string = connection_string[len('file://'):]
            con_params = urlparse.urlparse(f'{cls.SQL_DATABASE_SCHEME}://{file_connection_string}')
```
snippet you stumbled upon just above.

But these quirks and work-arounds are irritating me a bit, and a major part of the reason I want to try to re-implement most of the path/parsing stuff at some point. I will add a more in-depth post regarding my thoughts about that later

nickeopti commented on terracotta/drivers/relational_base.py:
```diff
@@ -0,0 +1,486 @@
+        self._CONNECTION_PARAMETERS = self._parse_connection_string(path)
+        cp = self._CONNECTION_PARAMETERS
+        resolved_path = self._resolve_path(cp.path[1:])  # Remove the leading '/'
+        connection_string = f'{cp.scheme}+{self.SQL_DRIVER_TYPE}://{cp.netloc}/{resolved_path}'
```
You may a point there. Also in the context of the previous couple remarks above this. I'll add an explainer later

nickeopti commented on terracotta/drivers/relational_base.py:
```diff
@@ -0,0 +1,486 @@
+                    )
+                return connection
+            try:
+                with _connect_with_exceptions_converted() as connection:
```
Thanks, that's much better! I've adjusted it to
```python
    @contextlib.contextmanager
    def connect(self, verify: bool = True) -> Iterator:
        @convert_exceptions(_ERROR_ON_CONNECT, sqla.exc.OperationalError)
        def get_connection() -> Connection:
            return self.sqla_engine.connect().execution_options(isolation_level='READ UNCOMMITTED')

        if not self.connected:
            try:
                with get_connection() as connection:
                    self.connection = connection
                    self.connected = True
```

That fancy `with ()` statement appears to be a Python 3.9 or greater thing.

nickeopti commented on terracotta/drivers/sqlite.py:
```diff
@@ -94,299 +59,18 @@ def __init__(self, path: Union[str, Path]) -> None:
-        super().__init__(os.path.realpath(path))
+    @classmethod
+    def _resolve_path(cls, path: str) -> str:
+        return os.path.realpath(Path(path).resolve())
```
That is a combination of `os.path.realpath(path)` from https://github.com/DHI-GRAS/terracotta/blob/main/terracotta/drivers/sqlite.py#L110 and `Path(url_or_path).resolve()` from https://github.com/DHI-GRAS/terracotta/blob/main/terracotta/drivers/__init__.py#L80 😉 
Do you have any preference of which to use?

dionhaefner replied:
Either is fine. I personally like `os.path` better for this use case if you need a tie breaker.

nickeopti commented:
Thanks for the suggested improvements. I think this has helped a lot! Status update:

> > All major functionality is preserved -- however, there are some smaller functional changes:
> > 
> > 1. Previous `sqlite` driver checked DB version compatibility every time a connection was established, while `mysql` driver only checked once per driver instance. This branch uses the `mysql` driver approach for all drivers.
> 
> OK
> 
> > 2. The drivers' `_connection_callback` is renamed to more descriptive `_verify_db_version`. This is slightly weird in the `sqlite_remote` driver, which actually uses that method as a generic callback, but I still believe it is more descriptive.
> 
> This is supposed to be a generic hook that just happens to be mostly about verifying the DB version. I prefer `_connection_callback`.
> 
Done.

> > 3. Several internal instance variables are renamed as well. If this may be troublesome for users depending on them, then lets rename them back again. But it is only done on instance variables, so I kinda hope no one relies on them.
> 
> This is fine.
> 
> > 4. Previous `mysql` driver cached keys, while the `sqlite` didn't. This branch uses `mysql` approach for all drivers.
> 
> I don't see any caching logic in any of the drivers?
>
Ah, I can see I was a bit ambiguous there. That sentence was regarding the `key_names` property.
 
> > 5. `settings.LAZY_LOADING_MAX_SHAPE` was used when lazy loading metadata in the previous `sqlite` driver, but not in the `mysql` driver. For simplicity, this branch uses the `mysql` approach (of ignoring `settings.LAZY_LOADING_MAX_SHAPE`) for all drivers. I honestly have no idea what that does, nor why it was implemented only in the one driver. Please tell me about it, if it is relevant :D
> 
> Yes, this is very much relevant and should be used. The fact that this wasn't used in the MySQL driver was a bug.
> 
> Lazy loading means computing raster metadata the first time it is requested instead of during ingestion. But there's a dilemma: Computing metadata is costly, so if you have a huge raster the request would time out. Instead we compute the dataset on a smaller version of big rasters (downsampled to `LAZY_LOADING_MAX_SHAPE`) to trade accuracy for speed. So lazy loading is about quick-and-dirty data visualization, not production quality services.
> 
Good to know! That is now implemented.

> > 6. `terracotta.get_driver()` now caches drivers only per process, not globally. This is a necessity due to how the SQLAlchemy engine is created on driver instance instantiation, whereas the previous code only created connections when they were about the be used. Otherwise the same engine could end up being shared across processes, which resulted in some curious errors.
> 
> OK
> 
> > 7. The `key_names` table is created with an additional `index` column, which is used to ensure that keys are retrieved from the database in the correct order. **This is vital, and currently a breaking change**. We need to decide how to handle this. This references [Internal reliance on physical order of DB rows #241](https://github.com/DHI-GRAS/terracotta/issues/241).
> 
> Ideally we would handle this via a migration script, see #245. If you're not tired of working on Terracotta yet it would be great if you could handle this :) Otherwise we'll have to accept that people will have to re-create their databases.
> 
I'd like to do that (at some point), but I may have to prioritise other projects first.

> > 8. The `mysql` no longer has a `MySQLCredentials` object to store details. Instead, this information is used straight-away to create the connection string.
> 
> OK
> 
It now actually has gotten a `self.url: sqlalchemy.engine.URL` object storing such information.

> > 9. The `RemoteSQLiteDriver` now has a `_local_path` field, which corresponds to what was previously contained in the inherited `path` field. The `path` field now contains a connection string instead.
> 
> OK
> 
> > Also, this branch handles paths somewhat differently -- particularly internally. This is done in an attempt to make the handling of paths and particularly connection strings as database-vendor agnostic as possible. However, I've also tried to retain the current functionality and conventions, and this has honestly lead to the entire path/connection string/connection parameters handling being somewhat clumsy. It works, so I guess everything is acceptable, but I really think we could improve that part of the code base significantly, _if_ we are ready to change the assumptions on how paths are stored, accessed, and normalized -- stemming all the way from the generic `Driver` class. However, that may belong to a future PR.
> 
> What specifically would you improve?
> 
I have rewritten the parsing part, and I think that has become much nicer now! I no longer wish to change anything there.

> > I have broken _one_ connection string input pattern, though. Connection strings of the form `test.example.com:3306/foo` does not currently work. Paths like
> > ```
> > * `test.example.com/foo`
> > 
> > * `user@test.example.com/foo`
> > 
> > * `user:password@test.example.com/foo`
> > 
> > * `mysql://test.example.com:3306/foo`
> > 
> > * `mysql://user:password@test.example.com/foo`
> > 
> > * `mysql://user:password@test.example.com:3306/foo`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > all work, but specifying a `port` without specifying a `scheme` does not. In fact, this trips up `urllib.parse.urlparse`, which this branch relies upon. We can of course implement some hard-coded checks for that particular path pattern -- but if it is alright with all of you, I think I would actually like to disallow that path pattern: if it is not parsed correctly by `urlparse`, shouldn't we consider it an invalid path, rather than try to fix the path? The failing `tests/drivers/test_drivers.py::test_normalize_url[mysql]` test covers exactly that question.
> 
> See suggestion above.

Fixed now, thanks for the solution.

> 
> > I have no idea what causes the `setup.py::mypy-status` and `terracotta/cache.py::mypy` tests to fail -- nor am I sure whether it has anything to do with these code changes.
> 
> It has nothing to do with your changes, just mypy getting stricter over time.
> 
> > BTW, even though SQLAlchemy supports thread-safe connection pools, this current implementation is _not_ thread-safe. As far as I can figure out, we would have to rewrite how the `requires_connection` decorator works quite substantially in order to achieve thread-safety. I propose this belongs to a future PR, if it is something we want to have.
> 
> Ultimately we would definitely want thread-safety. I've opened an issue to track this (#246).

nickeopti commented:
Regarding design decisions (just for some context):

My overarching design goal for this refactor was to eliminate as much code duplication as possible, while keeping the overall structure more or less intact. I really think the previous version (current `main` branch) suffered from code duplication -- both because it was hard to implement new stuff, and because the two drivers had become slightly fragmented in their functionality. 

So I really wanted to counter this as much as possible -- but I see I may have taken the duplication elimination a bit too far. I may have made the common code base (`RelationalDriver`) slightly too complicated in order to keep each driver as simple as possible.
As we (in hindsight) probably aren't going to be adding new drivers all the time, some extra complexity in the individual drivers are probably more acceptable than I had in mind 😄 

I think the `timeout` connection parameter setting, which you pointed out, is just on the edge of whether it ought to be moved into the individual drivers. But that is the only thing I'm considering moving.  For instance, I really like how all the path / connection string parsing is done centrally now (especially after the recent rewrite of that part).


Regarding structure:
 
> I'm a bit torn regarding the new structure of the code (a lot of this is inherited from the old version, so not really your fault). Particularly, I think the inheritance tree is way too deep: Driver -> RasterDriver -> RelationalDriver -> SQLiteDriver -> RemoteSQLiteDriver. If we can I would like to stick to a maximum of 3 levels.
> 
> I think we should remove the RasterDriver mixin (#159), but that is something we can do after handling this PR.
>
> Right now, I see 2 ways going forward:
> 
> 1. Remove some logic from RelationalDriver and move it to the subclasses to purge some class variables. Do not make RemoteSQLiteDriver inherit from SQLiteDriver, but from RelationalDriver instead. This comes at the cost of a bit more code duplication but should make the interface much cleaner.
> 2. Remove all subclasses (SQLiteDriver, MySQLDriver, RemoteSQLiteDriver) and handle everything inside RelationalDriver. I.e., driver = RelationalDriver(provider="sqlite", path="..."). This comes at the cost of increased complexity and some extra conditionals in the RelationalDriver` but also eliminates more inheritance.
> 
> I am leaning towards option 1 because I like how clean we can implement e.g. the remote SQLite driver (just override some methods and you're done), but what is your opinion?

I personally don't mind deep inheritance trees, as long as the inheritance is semantically logical, and that each level has clearly distinct purposes. And I think that is more or less the case here.

For instance, the `RemoteSQLiteDriver` is fundamentally a specialised `SQLiteDriver`, and hence it seems obvious to me that it should indeed inherit from it. This also makes any future changes/improvements to the `SQLiteDriver` automatically apply to the `RemoteSQLiteDriver`, as they probably should, without any code duplication (nor risk of functionality fragmentation).

> Remove some logic from RelationalDriver and move it to the subclasses to purge some class variables.

After the rewrite of the parsing logic, I honestly think the only real contender here is the `SQL_TIMEOUT_KEY` as mentioned above. I'm unsure whether that is worth it.

>  I am leaning towards option 1 because I like how clean we can implement e.g. the remote SQLite driver (just override some methods and you're done)

I fail to see how that would make the `RemoteSQLiteDriver` simpler, but please elaborate, if you think it would.

> I think we should remove the RasterDriver mixin (#159)

I don't really understand what you mean here. If you propose merging `Driver` and `RasterDriver`, then I'm really unsure what to think. On one hand, can we imagine ever having a driver that is not operating on rasters? If not, then a merger could make sense. On the other hand, having `Driver` as (what other OOP languages would call) an interface was really nice and helpful during the development of the `RelationalDriver`. I like uncluttered, easy-to-read interfaces 😄

If you meant something along the way of this proposal
```python
class RelationalDriver(Driver, ABC):
    ...
class MySQLDriver(RelationalDriver, RasterDriver):
    ...
```
then I'm also unsure what to think. I somehow actually kinda like it. It seems like such a neat compositional architecture, where you can swap either without affecting the other. I'm just not used to multiple inheritance (I had genuinely forgotten that Python even supports it), so I don't know the (dis-)advantages well enough.
If we're going that route, though, then we shall move the `_key_dict_to_sequence` method from the `RasterDriver` to the `Driver`.

dionhaefner commented:
The problems with complicated inheritance relationships I see are these:

1. It's less and less obvious to the programmer which method is being called the further you go down the inheritance tree. So if you call `self.foo()` you would in theory have to check *all* parent classes (in different files) to see whether they override this. In extreme cases this leads to something like the [deadly diamond of death](https://en.wikipedia.org/wiki/Multiple_inheritance#The_diamond_problem) where method resolution becomes ill-defined.
2. Forgetting a `super()` in your overwritten method has catastrophic consequences.
2. It's cumbersome to ensure that overwritten methods inherit the correct docstrings and type signatures from parent classes, so there's a risk that information becomes lost / obscured (can be alleviated with aggressive testing though).

None of these sins are unforgivable, but there's a reason why many have been falling out of love with OOP / inheritance. So if we can reduce the complexity without sacrificing too much it would be worth doing.

Some possible inheritance layouts:

```
1. Driver -> RasterDriver -> RelationalDriver -> SQLiteDriver -> RemoteSQLiteDriver
(current)

2. Driver -> RelationalDriver -> SQLiteDriver -> RemoteSQLiteDriver
(move raster logic in RasterDriver to functional module and import / use that in RelationalDriver)

3. RasterDriver -> RelationalDriver -> SQLiteDriver -> RemoteSQLiteDriver
(do we really need a base class if we only have 1 implementation?)

4. Driver -> RasterDriver -> RelationalDriver -> RemoteSQLiteDriver
(some more code deduplication by not inheriting from SQLiteDriver but also much less complexity)

5. Driver -> RasterDriver -> RelationalDriver
(all logic could be handled inside 1 class, see above)
```

I still like `Driver -> RelationalDriver -> RemoteSQLiteDriver`. Since the `Driver` class has *only* abstract members it doesn't increase complexity much (it's just an interface), and having database providers as subclasses makes it much cleaner to implement stuff like remote SQLite DBs (compared to handling it all inside `RelationalDriver`).

We could also loop in @mrpgraae as language lawyer / OOP police 😄

mrpgraae commented:
:police_officer:

While I see the beauty in using inheritance to eliminate code duplication while keeping a clean interface, I agree with @dionhaefner that long chains of inheritance can be cumbersome to work with, so it might be worth it to be a bit more pragmatic here.

Rather than pointing at any of yours, I'm just going to put an entirely new suggestion on the table :smile: 

Since loading rasters from some data source (disk, S3, etc.) is fundamentally orthogonal to the task of querying and inserting metadata from some database, it seems that we should instead have 2 separate class hierarchies. Examples:

```
MetadataDriver -> RelationalDriver -> SQLiteDriver
MetadataDriver -> RelationalDriver -> MySQLDriver
MetadataDriver -> BlockchainDriver (sorry)

RasterDriver -> S3Driver
RasterDriver -> FileDriver
```

Then use composition to put them together, i.e. a `MetadataDriver` takes a `RasterDriver` instance when it's initialized and store it as an internal variable.

Then `MetadataDriver`s know how to query datasets given some criteria and when they get metadata back, that metadata also contains a `path`. The `MetadataDriver` knows nothing about this path is (could be path to a file on disk, to S3, or something else) but it knows that it just needs to hand it over to the `RasterDriver` and it will get tiles back.

This way we could have any combination of SQLite / S3, SQLite / File, MySQL / S3, etc.

Now I have described what I think is the ideal way to do it, which I realize is not very helpful to the current PR, it's also been a while since I contributed to Terracotta, so I might have missed some obvious reason as to why this way would just not work. Forgive me, if that's the case.

**Of all the suggestions that @dionhaefner mentions above, I think his suggestion 2 is closest to what I suggest here, so I would go with that. I also agree with @nickeopti that the `SQLiteDriver -> RemoteSQLiteDriver` abstraction is just too sweet to get rid of for pragmatic reasons.**

Thank you for this big contribution @nickeopti, it's nice to see such high-quality contributions to the project!

mrpgraae commented:
instead of passing a `path` to the `RasterDriver`, the `MetadataDriver` should of course just pass the `keys`. Then the instance of `RasterDriver` will know how to turn that into a `path` on it's given backend.

mrpgraae commented:
Basically, I'm advocating that the orthogonality of the 2 problems calls for [**composition over inheritance**](https://en.wikipedia.org/wiki/Composition_over_inheritance)

dionhaefner commented:
Thanks for the suggestion. I think I would actually go 1 step further, and handle both tasks through composition. I.e.,

```python
driver = TerracottaDriver(
    metastore=SQLiteMetaStore("/foo/bar/x.sqlite"), 
    rasterstore=GeoTiffRasterStore()
)

# or, when / if we support storing everything in PostGIS

driver = TerracottaDriver(
    metastore=PostgresMetaStore("postgres://root:root@example.com"), 
    rasterstore=PostGISRasterStore("postgres://root:root@example.com")
)
```

with class hierarchy

```
TerracottaDriver

MetaStore -> RelationalMetaStore -> SQLiteMetaStore
MetaStore -> RelationalMetaStore -> MySQLMetaStore
MetaStore -> RelationalMetaStore -> RemoteSQLiteMetaStore
MetaStore -> RelationalMetaStore -> PostgresMetaStore

RasterStore -> GeoTiffRasterStore
RasterStore -> PostGISRasterStore
```

WDYT?

mrpgraae commented:
> Thanks for the suggestion. I think I would actually go 1 step further, and handle both tasks through composition. I.e.,
> 
> ```python
> driver = TerracottaDriver(
>     metastore=SQLiteMetaStore("/foo/bar/x.sqlite"), 
>     rasterstore=GeoTiffRasterStore()
> )
> 
> # or, when / if we support storing everything in PostGIS
> 
> driver = TerracottaDriver(
>     metastore=PostgresMetaStore("postgres://root:root@example.com"), 
>     rasterstore=PostGISRasterStore("postgres://root:root@example.com")
> )
> ```
> 
> with class hierarchy
> 
> ```
> TerracottaDriver
> 
> MetaStore -> RelationalMetaStore -> SQLiteMetaStore
> MetaStore -> RelationalMetaStore -> MySQLMetaStore
> MetaStore -> RelationalMetaStore -> RemoteSQLiteMetaStore
> MetaStore -> RelationalMetaStore -> PostgresMetaStore
> 
> RasterStore -> GeoTiffRasterStore
> RasterStore -> PostGISRasterStore
> ```
> 
> WDYT?

@dionhaefner Yes, that occurred to me as well after posting my other post on Friday :slightly_smiling_face: 
That would definitely be the proper way to do it!

dionhaefner commented on terracotta/drivers/base.py:
```diff
@@ -3,21 +3,29 @@
 
 
-def requires_connection(fun: Callable[..., T]) -> Callable[..., T]:
+def requires_connection(
```
I don't think mypy supports multiple output types. Let's leave it, not worth the hassle.

dionhaefner commented:
How do you like this suggestion @nickeopti ? (Don't worry about the implementation, we can merge this PR soon and then I can have a stab at it myself. This has been going on long enough.)

nickeopti commented:
Wow, those are some elegant solutions! That definitely seems to be the better way to go.

At a high abstraction level, I certainly prefer @dionhaefner's all-in proposal, but @mrpgraae's initial proposal may have some benefits in regards to a couple of implementation details: (in the lingo of @dionhaefner's post)

1. `MetaStore`s need `compute_metadata()` (from `RasterStore`) for automatically computing the metadata in `insert()` as well as for lazy loading in `get_metadata()`. This could be handled by either:
    A: `MetaStore`s having a reference to the `RasterStore`. This could be handled by the `TerracottaDriver`.
    B: `MetaStore`s having a `compute_metadata` callback, where the `TerracottaDriver` supplies the `RasterStore`s `compute_metadata()`.
2. We would have to either:
    A: Supply `TerracottaDriver` with wrapper methods for all the (abstract) methods from `MetaStore` (and `RasterStore`?), or
    B: Make a breaking API change, such that users need to call e.g. `driver.metastore.get_datasets()` instead of the current `driver.get_datasets()`.

Just some further design decisions to make 😄, in case we go all-in with the second proposal, whereas @mrpgraae's initial proposal completely bypasses those. But despite that, I quite like the all-in, fully compositional proposal with the `TerracottaDriver`.

> instead of passing a path to the RasterDriver, the MetadataDriver should of course just pass the keys. Then the instance of RasterDriver will know how to turn that into a path on it's given backend.

I think we still need to have the `MetaStore`/`MetadataDriver` to give a path, rather than just the keys. I think one of the neat features of the current version is that the database can make an abstraction between keys and physical file paths, so we probably shouldn't remove that. However, we may want to make it a more abstract `identifier` (or something along that way), which just happens to be a path in the case of a file-based `RasterDriver`/`GeoTiffRasterStore` -- but could for instance be a primary key in the case of a `PostGISRasterStore`.

> (Don't worry about the implementation, we can merge this PR soon and then I can have a stab at it myself. This has been going on long enough.)

I'm fine with spending the time to get this PR to a state we are happy with, before merging 😄 No need to hurry from my point of view.

dionhaefner commented:
Yes, the `RasterStore` has to be agnostic of keys, so @mrpgraae's final proposal doesn't quite work :) But it works out well with some small modifications.

The `insert` method could look something like this:

```python
class TerracottaDriver:
    ...

    def insert(self, handle, ...):
        mtd = self.raster_store.compute_metadata(handle)
        self.meta_store.insert(handle, metadata=mtd)
```

In this case, `handle` could be anything that unique identifies a raster. Currently, this would always be a file path, but in the future it cold also be a database key or so. The nice things is that `TerracottaDriver` doesn't need to know what `handle` is, it can just pass it on.

Lazy loading could look like this:

```python
mtd = self.meta_store.get_metadata(keys)

if mtd is None:
    # metadata is not computed yet, trigger lazy loading
    handle = self.meta_store.get_dataset(keys)
    mtd = self.raster_store.compute_metadata(handle)
    self.meta_store.insert(keys, handle, metadata=mtd)

return mtd
```

All interaction from the outside will happen with `TerracottaDriver`. Many of the methods will just be thin wrappers around `RasterStore` and `MetaStore` methods, but that's fine (and allows us to control exposed docstrings and signatures much more precisely, so I think it's a gain on all fronts).

This way, `RasterStore` and `MetaStore` are single-purpose and completely encapsulated / agnostic of each other. All orchestration and data exchange is happening through `TerracottaDriver`.

dionhaefner commented:
> I'm fine with spending the time to get this PR to a state we are happy with, before merging 😄 No need to hurry from my point of view.

That's great to hear. It's all yours, then!

mrpgraae commented:
Happy to hear that everyone likes the solution.

Couldn't quite remember if keys were strictly a metadata abstraction or not and was in a bit of a rush when I wrote the proposed solution 😄but being able to decouple keys from paths sound like a excellent concept that we should keep, so good idea to just pass a handle to the RasterDriver.

With regards to fixing the issues with TerracottaDriver, I also think it's perfectly fine to lift out the functionality that will be needed across both meta and raster drivers into thin methods in TerracottaDriver. I have seen that approach before in similar situations.

dionhaefner commented:
@nickeopti we good to go?

nickeopti commented on setup.py:
```diff
@@ -35,6 +35,7 @@
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: 3.8',
         'Programming Language :: Python :: 3.9',
+        'Programming Language :: Python :: 3.10',
```
Not really related, but is this intentional?

dionhaefner replied:
Yup

nickeopti replied:
Alrighty :D

nickeopti commented:
Yes, I believe this is mergeable!
